src/kvstore/Part.cpp

```diff
@@ -5,34 +5,210 @@
+
+    appendAsync(FLAGS_cluster_id, std::move(log))
+        .then([callback = std::move(cb)] (AppendLogResult res) {
+            callback(toResultCode(res));
+        });
+}
+
+
+void Part::asyncMultiPut(const std::vector<KV>& keyValues, KVCallback cb) {
+    std::string log = encodeMultiValues(OP_MULTI_PUT, keyValues);;
+
+    appendAsync(FLAGS_cluster_id, std::move(log))
+        .then([callback = std::move(cb)] (AppendLogResult res) {
+            callback(toResultCode(res));
+        });
+}
+
+
+void Part::asyncRemove(folly::StringPiece key, KVCallback cb) {
+    std::string log = encodeSingleValue(OP_REMOVE, key);;
+
+    appendAsync(FLAGS_cluster_id, std::move(log))
+        .then([callback = std::move(cb)] (AppendLogResult res) {
+            callback(toResultCode(res));
+        });
 }
 
-void SimplePart::asyncRemove(const std::string& key, KVCallback cb) {
-    CHECK_NOTNULL(engine_);
-    auto ret = engine_->remove(key);
-    cb(ret, HostAddr(0, 0));
+
+void Part::asyncMultiRemove(const std::vector<std::string>& keys, KVCallback cb) {
+    std::string log = encodeMultiValues(OP_MULTI_REMOVE, keys);;
+
+    appendAsync(FLAGS_cluster_id, std::move(log))
+        .then([callback = std::move(cb)] (AppendLogResult res) {
+            callback(toResultCode(res));
+        });
 }
 
-void SimplePart::asyncMultiRemove(std::vector<std::string> keys, KVCallback cb) {
-    CHECK_NOTNULL(engine_);
-    auto ret = engine_->multiRemove(std::move(keys));
-    cb(ret, HostAddr(0, 0));
+
+void Part::asyncRemovePrefix(folly::StringPiece prefix, KVCallback cb) {
+    std::string log = encodeSingleValue(OP_REMOVE_PREFIX, prefix);;
+
+    appendAsync(FLAGS_cluster_id, std::move(log))
+        .then([callback = std::move(cb)] (AppendLogResult res) {
+            callback(toResultCode(res));
+        });
 }
 
-void SimplePart::asyncRemoveRange(const std::string& start,
-                                  const std::string& end,
-                                  KVCallback cb) {
-    CHECK_NOTNULL(engine_);
-    auto ret = engine_->removeRange(start, end);
-    cb(ret, HostAddr(0, 0));
+
+void Part::asyncRemoveRange(folly::StringPiece start,
+                            folly::StringPiece end,
+                            KVCallback cb) {
+    std::string log = encodeMultiValues(OP_REMOVE_RANGE, start, end);;
+
+    appendAsync(FLAGS_cluster_id, std::move(log))
+        .then([callback = std::move(cb)] (AppendLogResult res) {
+            callback(toResultCode(res));
+        });
+}
+
+
+void Part::onLostLeadership(TermID term) {
+    VLOG(1) << "Lost the leadership for the term " << term;
+}
+
+
+void Part::onElected(TermID term) {
+    VLOG(1) << "Being elected as the leader for the term " << term;
+}
+
+
+std::string Part::compareAndSet(const std::string& log) {
+    UNUSED(log);
+    LOG(FATAL) << "To be implemented";
+}
+
+
+bool Part::commitLogs(std::unique_ptr<LogIterator> iter) {
+    auto batch = engine_->startBatchWrite();
```

Comment:
Record the latest committedLogId

Reply:
Good catch 👍  This has not been implemented. I'll add it to the ```commit()```